wakuv2/missing_messages.go

```diff
@@ -0,0 +1,257 @@
+package wakuv2
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"slices"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/peer"
+	"google.golang.org/protobuf/proto"
+
+	"go.uber.org/zap"
+
+	gethcommon "github.com/ethereum/go-ethereum/common"
+
+	"github.com/waku-org/go-waku/waku/v2/protocol"
+	"github.com/waku-org/go-waku/waku/v2/protocol/store"
+
+	"github.com/status-im/status-go/logutils"
+	"github.com/status-im/status-go/wakuv2/common"
+
+	"github.com/waku-org/go-waku/waku/v2/protocol/pb"
+)
+
+const maxContentTopicsPerRequest = 10
+const maxAttemptsToRetrieveHistory = 3
+const delay = 10 * time.Second
+
+type TopicInterestPerStorenode struct {
+	peerID        peer.ID
+	pubsubTopic   string
+	contentTopics []string
+	lastChecked   time.Time
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func (p TopicInterestPerStorenode) Equals(other TopicInterestPerStorenode) bool {
+	if p.peerID != other.peerID {
+		return false
+	}
+
+	if p.pubsubTopic != other.pubsubTopic {
+		return false
+	}
+
+	slices.Sort(p.contentTopics)
+	slices.Sort(other.contentTopics)
+
+	for i, contentTopic := range p.contentTopics {
+		if contentTopic != other.contentTopics[i] {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (w *Waku) CheckForMissingMessages(peerID peer.ID, pubsubTopic string, contentTopics []string) error {
+	w.topicInterestMu.Lock()
+	defer w.topicInterestMu.Unlock()
+
+	ctx, cancel := context.WithCancel(w.ctx)
+	newMissingMessageRequest := TopicInterestPerStorenode{
+		peerID:        peerID,
+		pubsubTopic:   pubsubTopic,
+		contentTopics: contentTopics,
+		lastChecked:   w.timesource.Now().Add(delay),
+		ctx:           ctx,
+		cancel:        cancel,
+	}
+
+	currMessageVerificationRequest, ok := w.topicInterest[pubsubTopic]
+	if !ok || !currMessageVerificationRequest.Equals(newMissingMessageRequest) {
+		if ok {
+			currMessageVerificationRequest.cancel()
+		}
+
+		w.topicInterest[pubsubTopic] = newMissingMessageRequest
+	}
+
+	return nil
+}
+
+func (w *Waku) checkForMissingMessages() {
+	defer w.wg.Done()
+	defer w.logger.Debug("checkForMissingMessages - done")
+
+	t := time.NewTicker(time.Minute)
+	defer t.Stop()
+
+	w.topicInterest = make(map[string]TopicInterestPerStorenode)
+
+	var semaphore = make(chan struct{}, 3)
+	for {
+		select {
+		case <-t.C:
+			w.topicInterestMu.Lock()
+			for _, request := range w.topicInterest {
+				semaphore <- struct{}{}
```

Comment:
Wondering if this semaphore can cause a queue of requests if store node response is slow.

Inside each FetchHistory we are only considering 10 contentTopics to be fetched per request, and based on tests so far if a user is part of 2 communities and has some contacts atleast 160 odd filters are installed which means this would result in an underlying 16 requests with each having a retry and delay to it. If store nodes are slow, and each request would take few seconds which would mean 1 minute(ticker interval) would catchup quickly in 2-3 cycles.  That added with pageSize limitations of store-queries as well will include lot of network operations.

The semaphore idea keeps a limit on number of parallel requests, but i don't think this would scale well. Maybe these can be optimized as part of separate PR.

Few suggestions i can think of. Maybe while testing this freature you may get better way to handle this.
1. increase number of topics queried in a request. But still pageSize can cause a limitation here.
2. Filter out the filters that need not be fetched e.g: pushnotification filters and any other application specific filters that can be filtered out to reduce count.
3. Maybe parallelize requests further at lower layer. 
4. Split network requests(only these can be gaurded under semaphore) and processing of success responses(can be parallelize more).

Reply:
1. maybe we could start with `an INNER JOIN over a VALUES clause` which seems like a small change. cc @Ivansete-status 
3. Agreed, that would speed up if we have lot of queries to be made. Maybe the split can be done based on number of topics to query frequently.